src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).    EXPECT_EQ(dev->_summary.by_type(mtTest)->reserved(), 0);
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_NMT_NMTNATIVECALLSTACKSTORAGE_HPP
+#define SHARE_NMT_NMTNATIVECALLSTACKSTORAGE_HPP
+
+#include "utilities/growableArray.hpp"
+#include "utilities/nativeCallStack.hpp"
+
```

Comment:
Here, I would love it if we had smaller than pointer-sized callstack IDs. 

The simplest way would be to copy callstacks into a growable area, and use their numerical indices as ID. We can get by with 16 bits, or 32 bits if we think 64K callstacks are not enough (they should be enough).

Those can then be combined very efficiently with MEMFLAG and VMAState in the VMATree nodes.

The hashmap for reverse id lookup could be a standard hashmap of key=callstack, value=ID.

As a future possible improvement, that could then replace the MallocSiteTable which does almost the same thing. (only have to avoid using malloc then, because of recursivities).

Reply:
I still don't get how a growable array is supposed to work for open addressing while having the indices staying immutable :-).

How does this work:

```c++
  HTable ht; // Open-addressed GrowableArray with linear probing
  Index oldidx = ht.put(4);
  // A bunch of puts, leading to a resize of the array
  Index newidx = ht.put(4);
  assert(oldidx == newidx, "how is this ensured?");
```

How does this work?

Reply:
Ah, I get the confusion. This is not what I meant.

What I mean was:

At the moment you malloc space for NativeCallStack, then keep NativeCallStack* in the hash map. NativeCallStack* now uniquely identifies your stack.

What I meant is to place NativeCallStack in a growable array. Now, you have a 32-bit or even a 16-bit index into that array. That index uniquely identifies the stack. You keep that index the hashmap. The hashmap does not change. Hashmap storage has nothing to do with that array. This is not the bucket array.

Basically, you replace the malloc for the NativeCallStack with a placement-new in a new growable array. The rest stays the same. 

But now, you have a 32-bit or even 16-bit index, and that is smaller than a native pointer, which makes it possible to encode the stack information in a tree node much more succinctively. This makes it possible to encode the whole tree node metainfo very comfortably in a single 64-bit value. You can even get both in- and out-state of the VMATree into a single 64-bit value like this:

bits 0-7 MEMFLAGS in
bits 8-16 State in
bits 16-31 callstack index in

bits 32-39 MEMFLAGS out
bits 40-47 State out
bits 48-63 callstack index out

Reply:
Oh right of course, just store the NCS separately to the closed-addressing hashtable. I'm going for a 32-bit value just because that's the quickest. We can do a further compression round in a future PR.

If we really wanted to, we could also store the `Link`s in a GA and thus reduce their pointer sizes to 32-bits also. Still, future PR, IMHO.

Reply:
We're down to a 4-byte `StackIndex`, hooray! Resolving this conversation now, let's go further down this route in a future RFE.